src/Compilers/VisualBasic/Portable/Lowering/Instrumentation/DynamicAnalysisInjector.vb

```diff
@@ -0,0 +1,303 @@
+
+            ' The first point indicates entry into the method And has the span of the method definition.
+            _methodEntryInstrumentation = AddAnalysisPoint(MethodDeclarationIfAvailable(methodBody.Syntax), factory)
+        End Sub
+
+        Public Overrides Function CreateBlockPrologue(original As BoundBlock, ByRef synthesizedLocal As LocalSymbol) As BoundStatement
+            Dim previousPrologue As BoundStatement = MyBase.CreateBlockPrologue(original, synthesizedLocal)
+
+            If _methodBody Is original Then
+                _dynamicAnalysisSpans = _spansBuilder.ToImmutableAndFree()
+                ' In the future there will be multiple analysis kinds.
+                Const analysisKind As Integer = 0
+
+                Dim modulePayloadType As ArrayTypeSymbol = ArrayTypeSymbol.CreateVBArray(_payloadType, ImmutableArray(Of CustomModifier).Empty, 1, _factory.Compilation.Assembly)
+
+                ' Synthesize the initialization of the instrumentation payload array, using concurrency-safe code
+                '
+                ' Dim payload = PID.PayloadRootField[methodIndex]
+                ' If payload Is Nothing Then
+                '     payload = Instrumentation.CreatePayload(mvid, methodIndex, PID.PayloadRootField(methodIndex), payloadLength)
+                ' End If
+
+                Dim payloadInitialization As BoundStatement = _factory.Assignment(_factory.Local(_methodPayload, True), _factory.ArrayAccess(_factory.InstrumentationPayloadRoot(analysisKind, modulePayloadType, False), False, ImmutableArray.Create(_factory.MethodDefIndex(_method))))
+                Dim mvid As BoundExpression = _factory.ModuleVersionId(False)
+                Dim methodToken As BoundExpression = _factory.MethodDefIndex(_method)
+                Dim payloadSlot As BoundExpression = _factory.ArrayAccess(_factory.InstrumentationPayloadRoot(analysisKind, modulePayloadType, False), False, ImmutableArray.Create(_factory.MethodDefIndex(_method)))
+                Dim createPayloadCall As BoundStatement = _factory.Assignment(_factory.Local(_methodPayload, True), _factory.Call(Nothing, _createPayload, mvid, methodToken, payloadSlot, _factory.Literal(_dynamicAnalysisSpans.Length)))
+
+                Dim payloadNullTest As BoundExpression = _factory.Binary(BinaryOperatorKind.Equals, _factory.SpecialType(SpecialType.System_Boolean), _factory.Local(_methodPayload, False), _factory.Null(_payloadType))
+                Dim payloadIf As BoundStatement = _factory.If(payloadNullTest, createPayloadCall)
+
+                Debug.Assert(synthesizedLocal Is Nothing)
+                synthesizedLocal = _methodPayload
+
+                Return If(previousPrologue Is Nothing,
+                           _factory.StatementList(payloadInitialization, payloadIf, _methodEntryInstrumentation),
+                           _factory.StatementList(payloadInitialization, payloadIf, _methodEntryInstrumentation, previousPrologue))
+            End If
```

Comment:
The instrumentation for an individual statement should have the same span information as the statement, which requires a fresh factory per statement. The prologue instrumentation has the same span information as the method body.

I will rename the occurrences of "factory" to "methodBodyFactory", and add an explanatory comment to CollectDynamicAnalysis. (This applies to both languages.)